code for dictionay and deadlock detection

sureshdash2022-yb opened the pull request:
interview code assignments

sureshdash2022-yb committed:
code for dictionay and deadlock detection

interview code assignments

ashutosh-bapat commented:
Comment on Deadlock detection:
1. The algorithm doesn't seem to take into account the nodes. Also the program is not written for "You can assume max 1000 nodes and max 1000 queries on each node.". It assumes only what's mentioned in the problem statement.

ashutosh-bapat commented:
Comment on Dictionary
1. The data structure used to maintain the directory is a good choice.

But
1. the dictionary is not shared among processes as mentioned in the problem statement, "concurrent (by multiple processes)"
2. It does not take into account the limit "Maximum number of words in the dictionary is 1M." mentioned in the problem statement.

sureshdash2022-yb committed:
Review comments fix

Review comments:-

1. concurrent support
2.  support for 1M words for dictionary

sureshdash2022-yb commented:
Hi Ashutosh, 
fixed those review comments related to dictionary assignment:- 
1. concurrent support (the same process if creating multiple children then they can share the dictionary using shared memory implemented in mmap)
2. supporting 1 M words in the dictionary and throw an error when we insert more than that.

please review it and share with me if any gaps are still there.

Thanks
Suresh

sureshdash2022-yb commented:
Hi Ashutosh,
For the Assignment of "Deadlock detection":- 

just to clarify me. you want me to implement the parsing logic for nodes and each node's lock queue information right. 
after that those collected data need to be handover to my Implemented Banker's Algo, is understanding correct?

Note:- Need to take care of 1000 Nodes Max. and each node has a max of 1000 queries.


Thanks
Suresh

sureshdash2022-yb committed:
Review comments fix

1. Remove the file persist with mmap.

2. used tree lock for insert, search and delete operation

sureshdash2022-yb commented:
Hi Ashutosh, 
I have fixed the above review comments.

1. Remove the file persist with mmap.

2. used tree label lock for insert, search and delete operation 

Thanks
Suresh

ashutosh-bapat commented:
Thanks Suresh for improvements.

Locking the entire tree is going to hamper concurrency. Fine grained locks are expected.

sureshdash2022-yb committed:
Review comments fix

1. Fine grains the lock instead of taking lock whole over the dictionary.

sureshdash2022-yb committed:
Review comments fix

1. we have assumed that code will read json file and create appropriate object.

2. using the cluster information collecting object we validate the numbers Nodes max supported and
max query per node support.

ashutosh-bapat commented:
Hi Suresh,
If I understand correctly, you are still using the same lock but acquiring and releasing it more frequently. That won't help with concurrency.

sureshdash2022-yb committed:
Update dict.cpp

create an array of lock in each dictionary node level to optimize use of concurrency.
